internals/daemon/api_notices.go

```diff
@@ -0,0 +1,178 @@
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/canonical/x-go/strutil"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+// A very loose regex to ensure custom keys are in the form "domain.com/key"
+var customKeyRegexp = regexp.MustCompile(`^([a-z0-9-]+\.)+[a-z0-9-]+/[A-Za-z0-9./-]+$`)
+
+const (
+	maxNoticeKeyLength = 256
+	maxNoticeDataSize  = 4 * 1024
+)
+
+func v1GetNotices(c *Command, r *http.Request, _ *UserState) Response {
+	query := r.URL.Query()
+
+	typeStrs := strutil.MultiCommaSeparatedList(query["types"])
+	types := make([]state.NoticeType, 0, len(typeStrs))
+	for _, typeStr := range typeStrs {
+		noticeType := state.NoticeType(typeStr)
+		if !noticeType.Valid() {
+			return statusBadRequest("invalid notice type %q", typeStr)
+		}
+		types = append(types, noticeType)
+	}
+
+	keys := strutil.MultiCommaSeparatedList(query["keys"])
+
+	afterStr := query.Get("after")
+	var after time.Time
+	if afterStr != "" {
+		var err error
+		after, err = time.Parse(time.RFC3339, afterStr)
+		if err != nil {
+			return statusBadRequest("invalid after timestamp %q: %v", afterStr, err)
+		}
+	}
+
+	filter := &state.NoticeFilter{
+		Types: types,
+		Keys:  keys,
+		After: after,
+	}
+	var notices []*state.Notice
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	timeoutStr := query.Get("timeout")
+	if timeoutStr != "" {
+		// Wait up to timeout for notices matching given filter to occur
+		timeout, err := time.ParseDuration(timeoutStr)
+		if err != nil {
+			return statusBadRequest("invalid timeout %q: %v", timeoutStr, err)
+		}
+
+		ctx, cancel := context.WithTimeout(r.Context(), timeout)
+		defer cancel()
+
+		notices, err = st.WaitNotices(ctx, filter)
+		if errors.Is(err, context.Canceled) {
+			return statusBadRequest("request canceled")
+		}
+		// DeadlineExceeded will occur if timeout elapses; in that case return
+		// an empty list of notices, not an error.
+		if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+			return statusInternalError("cannot wait for notices: %s", err)
+		}
+	} else {
+		// No timeout given, fetch currently-available notices
+		notices = st.Notices(filter)
+	}
+
+	if len(notices) == 0 {
+		notices = []*state.Notice{} // avoid null result
+	}
+	return SyncResponse(notices)
+}
+
+func v1PostNotices(c *Command, r *http.Request, _ *UserState) Response {
```

Comment:
We should also talk about taking user into account here too.

Reply:
See above.